src/metabase/models/app.clj

```diff
@@ -9,7 +9,14 @@
 (models/defmodel App :app)
 
 ;;; You can read/write an App if you can read/write its Collection
-(derive App ::perms/use-parent-collection-perms)
+(doto App
+  (derive ::mi/read-policy.full-perms-for-perms-set)
+  (derive ::mi/write-policy.full-perms-for-perms-set))
+
+(defmethod mi/perms-objects-set App
+  [instance read-or-write]
+  (let [collection (db/select-one 'Collection :id (:collection_id instance))]
+    (mi/perms-objects-set collection read-or-write)))
```

Comment:
This is necessary because just using `::perms/use-parent-collection-perms` would lose namespace information and the "root collection check" hack wouldn't work.

Reply:
A few observations:

1.  You actually don't need to fetch the Collection to build the perms set -- Collection read and readwrite paths are just functions of the Collection ID (and namespace in this case if collection ID is `nil`). You should be able to use the functions in https://github.com/metabase/metabase/blob/master/src/metabase/models/permissions.clj#L397-L409 here. Always good to avoid app DB calls where we can
2. Since `:collection_id` is `nil` you should probably check it before doing `(db/select-one Collection :id nil)`, since that's always going to return `nil`... unfortunately Toucan 1 isn't smart enough to have us be able to tell it that `WHERE id = NULL` is going to return `NULL` in cases like this an optimize it out. So it's doing an unneeded app DB call in that case. I'd also wonder how `(mi/perms-objects-set nil :read)` is supposed to work -- I would have expected this will error.
3. It's better to avoid quoting model names like `'Collection` if you can -- that's only there as an escape hatch to prevent circular references between model namespaces. It does runtime resolution of the model which is a little more expensive than if it already has the var in place. It also means tools that build dependency graphs between namespaces can't get the full picture. So it's better to `(:require [metabase.models.collection :refer [Collection]])`

Reply:
`:collection_id` should never be `nil` in the current setup, but I'll reduce this case to checking the root permission in the `apps` namespace. I'll go with the proposal below and simply create the `Collection` instance we need.